### PR TITLE
basic first level permutation tests

### PR DESCRIPTION
**Pull request by MartinPerez**
Here I provide the localizer example modified to perform basic contrasts first level permutation tests. I also provide an initial proposal for a permutation API. I am convinced this API is far from optimal, particularly I find cumbersome the way I am passing the design matrix objs to the permutation test. I would support #40, since I think having a design matrix object would allow me to pass it and access all relevant information from it as I do from the FirstLevelGLM object.

I am not sure on what would be the appropriate way to write tests for this functionality. In any case first eye tests seem nice. Following we can see the comparison between the original z score map for the contrast reading-visual for uncorrected p<0.1, then the permuted version for uncorrected p<0.1 and finally the permuted version for corrected p<0.5. The correction is very strong. 600 permutations were performed.

reading-visual_z_map

reading-visual_unc_pval

reading-visual_cor_pval

### CODE REVIEW

**Comment by MartinPerez** [comment]
Cluster statistics would be easily included in line 112. I think it would be good to wait for #35 thresholding api to be confirmed to develop it.

**Comment by MartinPerez** [comment]
For second level permutation I think I could fetch the haxby dataset (6 subjects) and perform the test on the well known object vs faces contrast?

**Comment by bthirion** [comment]
For second-level analysis, I would take the localizer dataset and perform a one-sample test on a given contrast, then possibly a two-sample tests, e.g. where you compare two populations (e.g. male vs female) , or two contrasts (e.g. right motor versus left motor activity).

File: examples/plot_localizer_analysis_permutation.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+More specifically:
+
+1. A sequence of fMRI volumes are loaded
+2. A design matrix describing all the effects related to the data is computed
+3. a mask of the useful brain volume is computed
+4. A GLM is applied to the dataset (effect/covariance,
+   then contrast estimation)
+5. Permutation tests are performed to obtain uncorrected and corrected p values
+at each voxel for a contrast comparing only two conditions.
```
Comment (bthirion) [inline examples/plot_localizer_analysis_permutation.py]: 'comparing two conditions' (no 'only')

File: examples/plot_localizer_analysis_permutation.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+from nistats.glm import FirstLevelGLM
+from nistats.design_matrix import make_design_matrix
+from nistats import datasets
+from nistats.permutation_tests import first_level_permutation_test
+from nistats.utils import z_score
+import nibabel as nib
+
+
+n_proc = 6
```
Comment (bthirion) [inline examples/plot_localizer_analysis_permutation.py]: should be `n_jobs=1`
(examples should work with n_jobs=1, since people will often run them on a laptop).

File: examples/plot_localizer_analysis_permutation.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+#     "left-right": (contrasts["clicGaudio"] + contrasts["clicGvideo"]
+#                    - contrasts["clicDaudio"] - contrasts["clicDvideo"]),
+#     "H-V": contrasts["damier_H"] - contrasts["damier_V"],
+#     "audio-video": contrasts["audio"] - contrasts["video"],
+#     "video-audio": -contrasts["audio"] + contrasts["video"],
+#     "computation-sentences": (contrasts["computation"] -
+#                               contrasts["sentences"]),
+#     "reading-visual": contrasts["phrasevideo"] - contrasts["damier_H"]
+#     }
+
```
Comment (bthirion) [inline examples/plot_localizer_analysis_permutation.py]: You can safely remove lines 68-87.

File: examples/plot_localizer_analysis_permutation.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+contrasts = {
+    "reading-visual": contrasts["phrasevideo"] - contrasts["damier_H"]
+    }
+
+### Permutation tests ###########################################
+
+# Design matrix obj arrangement for paradigm permutations
+# What about creating design_matrix class to wrap make design matrix and
+# respective parameters. Then could just pass that object
+
```
Comment (bthirion) [inline examples/plot_localizer_analysis_permutation.py]: As far as I can see, you don't need an object here, because you are running a computation, and that there is nothing that you might further  re-use.

Comment (MartinPerez) [inline examples/plot_localizer_analysis_permutation.py]: What do you mean by not needing an object? that I could instead say contrasts['reading-visual'] = ....

File: nistats/permutation_tests.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+from sklearn.base import clone
+from nistats.design_matrix import make_design_matrix
+import time
+import sys
+
+
+def _meet_contrast_conditions(con_val):
+    """Check if permutation test is allowed for given contrast values."""
+    # No F tests
+    if type(con_val[0]) is list:
```
Comment (MartinPerez) [inline nistats/permutation_tests.py]: I think this check is wrong...

File: examples/plot_localizer_analysis_permutation.py (hunk 4 above)
Comment (bthirion) [inline examples/plot_localizer_analysis_permutation.py]: But then, what you need is simply a dictionary ?
Maybe we should discuss this more directly.

File: examples/plot_localizer_analysis_permutation.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    display = plotting.plot_stat_map(thresholded_z_map, display_mode='z',
+                                     threshold=0.5, title=contrast_id)
+    display.savefig(path.join(write_dir, '%s_unc_pval.png' % contrast_id))
+
+    threshold = (fmri_glm.masker_.transform(cor_pval)[0] < 0.5)
+    thresholded_z_map = fmri_glm.masker_.inverse_transform(z_map_data * threshold)
+    display = plotting.plot_stat_map(thresholded_z_map, display_mode='z',
+                                     threshold=0.5, title=contrast_id)
+    display.savefig(path.join(write_dir, '%s_cor_pval.png' % contrast_id))
```
Comment (bthirion) [inline examples/plot_localizer_analysis_permutation.py]: display.show() at the end

File: examples/plot_localizer_analysis_permutation.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+    nib.save(unc_pval, path.join(write_dir, '%s_unc_pval.nii' % contrast_id))
+    nib.save(cor_pval, path.join(write_dir, '%s_cor_pval.nii' % contrast_id))
+
+    threshold = (fmri_glm.masker_.transform(unc_pval)[0] < 0.1)
+    thresholded_z_map = fmri_glm.masker_.inverse_transform(z_map_data * threshold)
+
+    display = plotting.plot_stat_map(thresholded_z_map, display_mode='z',
+                                     threshold=0.5, title=contrast_id)
+    display.savefig(path.join(write_dir, '%s_unc_pval.png' % contrast_id))
```
Comment (bthirion) [inline examples/plot_localizer_analysis_permutation.py]: this is  a misnomer because the image displays z values, not p-values.

File: examples/plot_localizer_analysis_permutation.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+for index, (contrast_id, (unc_pval, cor_pval)) in enumerate(res.iteritems()):
+    z_map, = fmri_glm.transform(contrasts[contrast_id],
+                                contrast_name=contrast_id,
+                                output_z=True)
+    nib.save(z_map, path.join(write_dir, '%s_z_map.nii' % contrast_id))
+    z_map_data = fmri_glm.masker_.transform(z_map)[0]
+
+    display = plotting.plot_stat_map(z_map, display_mode='z',
+                                     threshold=z_score(0.1),
```
Comment (bthirion) [inline examples/plot_localizer_analysis_permutation.py]: you're a bit extreme in the choice of the threshold. For an uncorrected  p-value, rather take .001

File: examples/plot_localizer_analysis_permutation.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+                                     title=contrast_id)
```
Comment (bthirion) [inline examples/plot_localizer_analysis_permutation.py]: the titles should be different for the 3 figures

File: nistats/permutation_tests.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        return False
+    # Not with less or more than 2 conditions
+    if len(np.where(con_val != 0.0)[0]) != 2:
```
Comment (bthirion) [inline nistats/permutation_tests.py]: if np.sum(con_val != 0) != 2:

Comment (bthirion) [inline nistats/permutation_tests.py]: bzw, you're assuming a t contrast. This should be checked.

File: nistats/permutation_tests.py (hunk 5 above)
Comment (bthirion) [inline nistats/permutation_tests.py]: Yes, this should checked by other means. 
I think that the API would clearer if you were providing the identifiers (strings) of two columns of the design matrix on the difference of which you want to make inference.

File: nistats/permutation_tests.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+        return False
+    return True
+
+
+def _permutations_glm(original_stat, glm_ref, imgs, design_matrix_objs,
+                      con_val, con_id, con_labels, n_perm_chunk=10000,
+                      two_sided_test=True,
+                      random_state=None,
+                      n_perm=10000,
+                      thread_id=1,
```
Comment (bthirion) [inline nistats/permutation_tests.py]: what is this thread_id ?

File: nistats/permutation_tests.py [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+    if len(np.where(con_val != 0.0)[0]) != 2:
+        return False
+    return True
+
+
+def _permutations_glm(original_stat, glm_ref, imgs, design_matrix_objs,
+                      con_val, con_id, con_labels, n_perm_chunk=10000,
+                      two_sided_test=True,
+                      random_state=None,
+                      n_perm=10000,
```
Comment (bthirion) [inline nistats/permutation_tests.py]: n_perm and n_perm_chunks are redundant.

File: nistats/permutation_tests.py [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+                      thread_id=1,
+                      verbose=0):
+    """Massively univariate group analysis with permuted OLS on a data chunk.
+    To be used in a parallel computing context.
+    Parameters
+    ----------
+    design_matrix_objs: list of [time_frames, paradigm, kwargs dict].
+        frame_times is an array of shape (n_frames,) representing the timing
+        of the scans in seconds. Paradigm is a DataFrame instance with the
```
Comment (bthirion) [inline nistats/permutation_tests.py]: paradigm

File: nistats/permutation_tests.py [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+        description of the experimental paradigm. kwargs are any other
+        arguments that could be passed to the function make_design_matrix
+        defined in nistats.design_matrix.
+
+    original_stat : array-like, shape=(n_voxels)
```
Comment (bthirion) [inline nistats/permutation_tests.py]: Do you need to give this ? This should be computed inside the function

File: nistats/permutation_tests.py [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+
+    random_state : int or None,
+        Seed for random number generator, to have the same permutations
+        in each computing units.
+
+    Returns
+    -------
+    unc_ranks_parts: array-like, shape=(n_voxels)
+        Accumulated count of permuted stats over original stats,
+        to compute uncorrected p-values. (limited to this permutation chunk).
```
Comment (bthirion) [inline nistats/permutation_tests.py]: The notion of permutation chunk is unclear at that point.

File: nistats/permutation_tests.py [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+        If False, only positive effects are considered as relevant. The null
+        hypothesis is that the effect is zero or negative.
+
+    random_state : int or None,
+        Seed for random number generator, to have the same permutations
+        in each computing units.
+
+    Returns
+    -------
+    unc_ranks_parts: array-like, shape=(n_voxels)
```
Comment (bthirion) [inline nistats/permutation_tests.py]: 'uncorrected_ranks_part'
Could you use a more explicit name ?

File: nistats/permutation_tests.py [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+        Accumulated count of permuted stats over original stats,
+        to compute uncorrected p-values. (limited to this permutation chunk).
+
+    smax_parts : array-like, shape=(n_perm_chunk, )
```
Comment (bthirion) [inline nistats/permutation_tests.py]: maybe `t_max_chunks`

File: nistats/permutation_tests.py [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+            perm_labels = rng.permutation(sel_labels['name'].tolist())
+            for lidx, perm_label in zip(labels_idx, perm_labels):
+                perm_paradigm.loc[lidx, 'name'] = perm_label
+
+            design_matrices.append(make_design_matrix(frame_times,
+                                                      perm_paradigm,
+                                                      **kwargs))
+
+        # Get permuted stat
+        glm_obj = clone(glm_ref)
```
Comment (bthirion) [inline nistats/permutation_tests.py]: You should probably not clone.

File: nistats/permutation_tests.py [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+                    "Job #%d, processed %d/%d permutations for contrast %s "
+                    "(%0.2f%%, %i seconds remaining)%s"
+                    % (thread_id, i, n_perm_chunk, con_id, percent, remaining,
+                       crlf))
+
+    return unc_rank_parts, smax_parts
+
+
+def first_level_permutation_test(contrasts, glm_ref, imgs,
+                                 design_matrix_objs, n_perm=50,
```
Comment (bthirion) [inline nistats/permutation_tests.py]: `n_perm=50` is not  a reasonable default for a number of permutations. I would take 1000

File: nistats/permutation_tests.py [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+        permuted_stat, = glm_obj.fit(imgs, design_matrices) \
+                                .transform(con_val, contrast_name=con_id,
+                                           output_z=False, output_stat=True)
+        permuted_stat = glm_obj.masker_.transform(permuted_stat)[0]
+
+        # For uncorrected p-values
+        if two_sided_test:
+            unc_rank_parts += (np.fabs(permuted_stat) < original_stat)
+        else:
+            unc_rank_parts += (permuted_stat < original_stat)
```
Comment (bthirion) [inline nistats/permutation_tests.py]: I thinks that storing the permutation values instead of the result of the threshold would avoid requiring the original stat in this function and would make it easier to understand: it would simply be a generator of statistical values to compare with an unknwon reference. The test would be handled in another function.
Try to keep all the functions as specific and simple as possible.

File: nistats/permutation_tests.py [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+    conditions = np.where(con_val != 0.0)
+    t0 = time.time()
+    for i in range(n_perm_chunk):
+        # Shuffle labels of the conditions in the design matrix
+        design_matrices = []
+
+        for frame_times, paradigm, kwargs in design_matrix_objs:
+            perm_paradigm = paradigm.copy()
+            con_idx = np.where(con_val != 0.0)[0]
+            conditions = [con_labels[x] for x in con_idx]
```
Comment (bthirion) [inline nistats/permutation_tests.py]: Wouldn't it make more sense to provide those as arguments, rather than con_val ?

File: nistats/permutation_tests.py [hunk 22]
```diff
@@ -0,0 +1,296 @@
...
+            selection = np.array([False for x in range(len(paradigm))])
+            for condition in conditions:
+                selection = selection | (perm_paradigm['name'] == condition)
```
Comment (bthirion) [inline nistats/permutation_tests.py]: `selection = [x in conditions for x in perm_paradigm['name']]` is clearer to me

File: nistats/permutation_tests.py (hunk 21 above)
Comment (bthirion) [inline nistats/permutation_tests.py]: Note that providing a list of identifiers rather than a contrast would actuallay make it possible to permute more than 2 items.

File: nistats/permutation_tests.py [hunk 23]
```diff
@@ -0,0 +1,296 @@
...
+            warnings.warn('Contrast %s do not meet conditions.'
+                          'Computation ignored.' % (con_id,))
+            continue
+
+        # Get original stat
+        design_matrices = []
+        for time_frames, paradigm, kwargs in design_matrix_objs:
+            design_matrices.append(make_design_matrix(time_frames, paradigm,
+                                                      **kwargs))
+        glm_obj = clone(glm_ref)
```
Comment (bthirion) [inline nistats/permutation_tests.py]: Cloning should be avoided

File: nistats/permutation_tests.py [hunk 24]
```diff
@@ -0,0 +1,296 @@
...
+                          'set to %s. This is incompatible so only %d jobs will '
+                          'be running. You may want to perform more permutations '
+                          'in order to take the most of the available computing '
+                          'resources.' % (n_perm, n_jobs, n_perm))
+            n_perm_chunks = np.ones(n_perm, dtype=int)
+
+        # initialize the seed of the random generator
+        rng = check_random_state(random_state)
+
+        per = joblib.Parallel(n_jobs=n_jobs, verbose=verbose)(
```
Comment (bthirion) [inline nistats/permutation_tests.py]: `permutation_results` rather than `per`
The names should always be as explicit as possible.

File: nistats/permutation_tests.py [hunk 25]
```diff
@@ -0,0 +1,296 @@
...
+              joblib.delayed(_permutations_glm)(
+                  original_stat, glm_ref, imgs, design_matrix_objs,
+                  con_val, con_id, design_matrices[0].columns.tolist(),
+                  n_perm_chunk=n_perm_chunk,
+                  two_sided_test=two_sided_test,
+                  random_state=rng.random_integers(np.iinfo(np.int32).max - 1),
+                  n_perm=n_perm, thread_id=thread_id, verbose=verbose)
+              for thread_id, n_perm_chunk in enumerate(n_perm_chunks))
+
+        unc_ranks_parts, smax_parts = zip(*per)
```
Comment (bthirion) [inline nistats/permutation_tests.py]: Same here regarding naming conventions.

**Comment by bthirion** [comment]
Please add tests. This will help to detect potential bugs.

**Comment by bthirion** [comment]
I'm going to look where the time is spent, as this business is rather slow.

**Comment by bthirion** [comment]
A huge amount of time is spent masking the data. This should be done once and for all. Hence the permutations should directly performed on the masked data. Contrast computation also has a large overhead.

**Comment by MartinPerez** [comment]
ok I will address all

**Comment by MartinPerez** [comment]
this is now completely off the new objects and API and out of the roadmap. So I would close it if you agree @bthirion.
